### ISSUE TITLE
Feature request: run the "source.organizeImports" code action as a before-save-hook

### ISSUE THREAD

**Issue by bcmills** [joaotavora/eglot#574]
(This feature request is forked from #411. A robust solution to #411 might address this request as well.)

The Go community has a standard tool, [`goimports`](https://pkg.go.dev/golang.org/x/tools/cmd/goimports), that automatically formats code and also automatically completes imports. It is widely used and recommended as a save hook. (For example, search for `gofmt-before-save` at https://github.com/dominikh/go-mode.el#features.)

The Go Language Server, `gopls`, provides the functionality of `goimports` as its `source.organizeImports` code action.

**I would like to be able to configure `eglot` to run the `source.organizeImports` action on save analogous to how I could configure `go-mode` to run `goimports`.**

----

`lsp-mode` provides this feature in the form of the [`lsp-organize-imports` function](https://github.com/emacs-lsp/lsp-mode/blob/c7dc0ed22a3fc3c759809b7e9d8a62be2311b31c/lsp-mode.el#L4490-L4497), which has the correct signature (and apparently the correct semantics) to be used as a `before-save-hook`.

#411 requests a similar function as a keybinding. https://github.com/joaotavora/eglot/issues/411#issuecomment-730474005 sketches out a possible implementation. However, due to the race described in https://github.com/joaotavora/eglot/issues/411#issuecomment-749305401, that implementation is not reliable enough to be a satisfactory save hook. (It works more-or-less well enough for an explicit keybinding, but only because the user can fairly easily wait for `eglot` and `gopls` to settle before pressing the key.)

CC @nemethf

**Comment by bcmills** [comment]
This is what I've got in the meantime as a workaround, based on https://github.com/joaotavora/eglot/issues/411#issuecomment-730474005 with an additional workaround for https://github.com/joaotavora/eglot/issues/411#issuecomment-749305401.

```elisp
;; eglot-organize-imports is hopefully a temporary stopgap until
;; https://github.com/joaotavora/eglot/issues/574 is addressed.
(defun eglot-organize-imports ()
  "Offer to execute the source.organizeImports code action."
  (interactive)
  (unless (eglot--server-capable :codeActionProvider)
    (eglot--error "Server can't execute code actions!"))
  (let* ((server (eglot--current-server-or-lose))
         (actions (jsonrpc-request
                   server
                   :textDocument/codeAction
                   (list :textDocument (eglot--TextDocumentIdentifier))))
         (action (cl-find-if
                  (jsonrpc-lambda (&key kind &allow-other-keys)
                    (string-equal kind "source.organizeImports" ))
                  actions)))
    (when action
      (eglot--dcase action
        (((Command) command arguments)
          (eglot-execute-command server (intern command) arguments))
        (((CodeAction) edit command)
          (when edit (eglot--apply-workspace-edit edit))
          (when command
            (eglot--dbind ((Command) command arguments) command
              (eglot-execute-command server (intern command) arguments))))))))

(defun eglot-organize-imports-on-save ()
  (defun eglot-organize-imports-nosignal ()
    "Run eglot-organize-imports, but demote errors to messages."
    ;; Demote errors to work around
    ;; https://github.com/joaotavora/eglot/issues/411#issuecomment-749305401
    ;; so that we do not prevent subsequent save hooks from running
    ;; if we encounter a spurious error.
    (with-demoted-errors "Error: %s" (eglot-organize-imports)))
  (add-hook 'before-save-hook #'eglot-organize-imports-on-save))

(add-hook 'go-mode-hook #'eglot-organize-imports-on-save)
```

**Comment by mdbergmann** [comment]
From my point of view, I wouldn't want to run this 'on save', not implicitly.
Wouldn't it be enough to wire it together in Emacs config?

**Comment by bcmills** [comment]
@mdbergmann, a function suitable for the user to wire into a save hook would satisfy my request, yes.

**Comment by vconcat** [comment]
#411 has closed.
So below is my updated config:
```
(defun +eglot-organize-imports() (call-interactively 'eglot-code-action-organize-imports))
(add-hook 'before-save-hook '+eglot-organize-imports nil t)
```

**Comment by bcmills** [comment]
@vconcat, that fails for me due to #608. 😞

**Comment by bcmills** [comment]
Ok, after #606 now I'm back to the failure mode described in https://github.com/joaotavora/eglot/issues/411#issuecomment-749305401.

I can at least simplify the wrapper now, but I still need a workaround for the synchronization error.

**Comment by joaotavora** [comment]
@bcmills . Please open a new issue for this.  Perhaps no need to make that big repro recipe (unless you can do it fast).  A transcript of the Eglot -> gopls conversation might be enough.

**Comment by bcmills** [comment]
Should be pretty easy to do on top of the my existing repo. New issue forthcoming.

**Comment by joaotavora** [comment]
Make sure you make the event transcript a visible thing.  I want to look at that first, then only perhaps download your repo and do all that.

**Comment by muffinmad** [comment]
@bcmills Do you know why server mark that code action as disabled?

```
server-reply (id:3) Mon Dec 21 21:32:32 2020:
(:jsonrpc "2.0" :result
	  [(:title "Organize Imports" :kind "source.organizeImports" :disabled
		   (:reason "")
		   :edit
		   (:documentChanges
		    [(:textDocument
		      (:version 0 :uri "file:///tmp/tmp.80ClKpYQ1o/main.go")
		      :edits
		      [(:range
			(:start
			 (:line 1 :character 0)
			 :end
			 (:line 1 :character 0))
			:newText "\nimport \"golang.org/x/mod/module\"\n")])]))]
	  :id 3)
```

Maybe Eglot should filter out disabled actions.

**Comment by joaotavora** [comment]
Disabled or not, does the action still make sense at the time it is presented to Eglot?  Does adding `import "golangblabla"` make sense at that moment as an "import organizing " action?

**Comment by bcmills** [comment]
Filed the version skew issue as #609.

**Comment by bcmills** [comment]
@muffinmad, I don't know why the action would be disabled at that point. (@stamblerre is probably the person to ask about that.)

**Comment by stamblerre** [comment]
Thank you so much for catching this! It will be fixed in the next `gopls` release: https://golang.org/cl/287796.

**Comment by wsw0108** [comment]
@bcmills I also see this ”disabled” lsp log，but gopls/eglot actually remove unused imports for me.

**Comment by muffinmad** [comment]
@wsw0108 Eglot at the moment doesn't filter out disabled actions and there will be no more disabled actions from `gopls` according to https://github.com/joaotavora/eglot/issues/574#issuecomment-769954286

**Comment by vconcat** [comment]
> @vconcat, that fails for me due to #608. 😞

Sorry that I forgot to mention #606 .

**Comment by wsw0108** [comment]
> @wsw0108 Eglot at the moment doesn't filter out disabled actions and there will be no more disabled actions from `gopls` according to [#574 (comment)](https://github.com/joaotavora/eglot/issues/574#issuecomment-769954286)

I see. That's the way. Thanks for point that.

**Comment by muffinmad** [comment]
@bcmills Can we close this feature request or something else must be done here?

**Comment by bcmills** [comment]
@muffinmad, I think the `eglot-code-action-organize-imports` function satisfies my feature request, and #609 seems to have addressed the remaining issues w.r.t. use in a save hook, so closing. Thanks!

**Comment by mekeor** [comment]
for the record, this works for me, based on the snippet from https://github.com/joaotavora/eglot/issues/574#issuecomment-769645691:

```elisp
  (defun own/eglot-organize-imports ()
    (call-interactively 'eglot-code-action-organize-imports))
  (defun own/before-saving-go ()
    ;; you might also like:
    ;; (add-hook 'before-save-hook #'eglot-format-buffer -10 t)
    (add-hook 'before-save-hook #'own/eglot-organize-imports nil t))
  (add-hook 'go-mode-hook #'own/before-saving-go)
```

**Comment by s-kostyaev** [comment]
Working for emacs 29:
```elisp
 (defun my-eglot-organize-imports () (interactive)
	 (eglot-code-actions nil nil "source.organizeImports" t))
  (add-hook 'before-save-hook 'my-eglot-organize-imports nil t)
  (add-hook 'before-save-hook 'eglot-format-buffer)
 ```
